### ISSUE TITLE
allow specifying singular value threshold for pinv

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/LinearAlgebra.jl#59]
It would be nice to have a way of controlling the `pinv` threshold for considering a singular value to be significant. I'm not sure what the best interface to that is, however, since you may want to just pass a value, but you may also want it to be a function of the full set of singular values.

Also: what's the rationale for the the `maximum(m,n)` factor in the current threshold value?

cc: @andreasnoack, @jiahao

**Comment by StefanKarpinski** [comment]
Maybe something like this:

``` diff
diff --git a/base/linalg/dense.jl b/base/linalg/dense.jl
index c2863a3..985a6b6 100644
--- a/base/linalg/dense.jl
+++ b/base/linalg/dense.jl
@@ -447,12 +447,13 @@ function (\){T<:BlasFloat}(A::StridedMatrix{T}, B::StridedVecOrMat{T})
 end

 ## Moore-Penrose inverse
-function pinv{T<:BlasFloat}(A::StridedMatrix{T})
+function pinv{T<:BlasFloat}(A::StridedMatrix{T};
+    thresh::Function = vals->eps(real(one(T)))*maximum(size(A))*maximum(vals))
     m, n = size(A)
     (m == 0 || n == 0) && return Array(T, n, m)
     SVD         = svdfact(A, true)
     Sinv        = zeros(T, length(SVD[:S]))
-    index       = SVD[:S] .> eps(real(one(T)))*max(m,n)*maximum(SVD[:S])
+    index       = SVD[:S] .> thresh(SVD[:S])
     Sinv[index] = 1.0 ./ SVD[:S][index]
     return SVD[:Vt]'scale(Sinv, SVD[:U]')
 end
```

**Comment by jiahao** [comment]
I'm surprised we don't allow the condition number threshold to be specified already. (LAPACK doesn't specify the lowest singular value explicitly, but rather the condition number = ratio of largest / smallest singular values, which I believe is the actual parameter that describes the accuracy of the computation.)

I think specifying `thresh` as a `Function` would be confusing to most people, since they would expect it to be a numeric cutoff.

**Comment by jiahao** [comment]
~~`max(m,n)` appears to be controlling for the case of rectangular matrices, where the thin SVD factorization doesn't return all the zero singular values but you want to account for them in constructing the pseudoinverse. /guess~~

It looks like the threshold is `eps()*max(m,n)`. Also, the singular values are conventionally ordered from largest to smallest, so `maximum(SVD[:S])` should be replaceable by `SVD[:S][1]`.

**Comment by StefanKarpinski** [comment]
I'd be happy with that, I just didn't know how best to express it. Hence the issue instead of just a pull request.

**Comment by jiahao** [comment]
Specifying the threshold as a numeric `cond` (condition number estimate) is probably the most conventional. (LAPACK uses the reciprocal `rcond`, but IIRC we uniformly avoid this in our linear algebra interface in favor of `cond`.)

**Comment by andreasnoack** [comment]
I am the author of `pinv` and maybe also the one who decided to use `cond` instead of `rcond` so the design decisions shouldn't be given too much weight. When I wrote `pinv` I didn't know a good criterion for the rank cutoff but could read on [wikpedia](http://en.wikipedia.org/wiki/Pinv#Singular_value_decomposition_.28SVD.29) that Numpy and MATLAB used the criterion above and thought it was sane.

**Comment by jiahao** [comment]
The Good Book (Golub and Van Loan) suggests that everyone is fine with just the condition number threshold. It's useful to have in practice since the pseudoinverse is notoriously sensitive to small singular values, and regularization by raising the condition number threshold is not uncommon, e.g. [(Hansen 1987)](http://link.springer.com/article/10.1007/BF01937276).

**Comment by timholy** [comment]
There's at least [one place](https://github.com/JuliaLang/julia/blob/master/base/linalg/factorization.jl#L576) where the threshold is not very `T`-aware (it's always `sqrt(eps(Float64))`.

**Comment by jiahao** [comment]
Ooh good catch. But this raises a potential problem, since we would want to be consistent in allowing the `cond` threshold to be specified consistently between least-squares `\` and `pinv`, but I don't see how it would be possible to specify the optional parameter in infix notation (something like `A\B,cond=1e-8`?), although `\(A, B, cond=1e-8)` would be perfectly fine.

**Comment by andreasnoack** [comment]
It is already there. See `\{T<:BlasFloat}(A::QRPivoted{T}, B::StridedMatrix{T}, rcond::Real)` around line 317 in `factorizations.jl`. The criterion here is the condition number but the default threshold is again somewhat arbitrary. We should definitely try to be consistent.

**Comment by jiahao** [comment]
It occurs to me that there are at least two ways to specify the threshold, namely the condition number cutoff _or_ the desired rank of the resulting approximation. So we may want to have two parameters.

Regarding the condition number cutoff, we should decide on:
1. whether to use `rcond` or `cond`. Currently we have `cond()` which computes the condition number, but the only functions that actually take a condition number parameter use `rcond`. (`cond` is also a parameter to `sqrtm()` but is  not a condition number, but instead determines the amount of output. `lapack.jl` also uses `rcond` uniformly thanks to LAPACK's convention.)
2. A default value. `rcond=sqrt(eps(T))` seems to be more popular than `eps(real(one(T)))*max(m,n)` in the existing codebase. However, I think it makes more sense of have as small an amount of regularization as we can get away with.

I read Golub and Van Loan more carefully and they actually outline two types of error analysis, depending on whether the emphasis is on the accuracy of the computed solution or minimizing the norm of the residual.

Note that the error in the computed solution from SVD solution is the sum of two terms:
1. The truncation error goes roughly as the root mean squared sum of the reciprocal singular values that were neglected.
2. The roundoff error is inversely proportional to the smallest reciprocal singular value _σ[r]_ that was retained.

Alternatively, the norm of the residual contains a term that proportional to the estimated rank _r_ times the regularized condition number _σ[1]/σ[r]_ (and this is the only term that depends on the condition number).
[(Golub and Van Loan 4/e, p. 291)](http://books.google.com/books?id=X5YfsuCWpxMC&lpg=PP1&pg=PA291#v=onepage&q&f=false)

So it would seem that `cond` or `rcond` makes sense for controlling the error on the residual, while specifying the rank cutoff `r` makes more sense if the accuracy of the solution is important.

**Comment by jiahao** [comment]
Implemented in JuliaLang/julia#9008
